### DISCUSSION TITLE
Public API to detect whether io_uring is usable

### DISCUSSION THREAD

**Discussion by cole-miller** [libuv/libuv#4034] [Ideas]
The project I'm working on (https://github.com/canonical/raft) needs to integrate some filesystem operations into its use of a libuv event loop. It was determined that running these operations on the thread pool, as uv_fs formerly did unconditionally, did not meet our performance requirements, and so we currently have a bespoke async file I/O subsystem based on KAIO (io_submit(2) and friends).

Now that uv_fs can use io_uring under the hood when the kernel supports it, we'd like to add a codepath that uses these interfaces, but only when the io_uring implementation will be used. If uv_fs is going to fall back to the thread pool, we'd like to keep using our KAIO implementation to avoid a performance regression. There's not currently a way to ask libuv whether io_uring was successfully set up -- would you consider adding such an API (presumably a "method" on uv_loop_t)? It would be possible to do a redundant check ourselves, but we'd want to copy exactly what libuv does, and that seems brittle.

**Comment by bnoordhuis** [comment]
I don't think such an API would help you because it's not a static yes/no.

Libuv decides per file operation whether to use io_uring or not, and that's fundamentally unknowable upfront. Among other things it depends on room in the (polled by the kernel) submit queue.

What could work is a flag that tells libuv to return an error instead of falling back to the thread pool. I don't want to add that straight away for a single downstream user but I'd be happy to help you make the necessary changes if you're willing to carry a local patch.

**Reply by cole-miller** [reply]
Thanks, that makes sense! When I was looking at the code I only saw the initial configuration (uv__iou_init) and missed the per-operation fallback.

I think a per-operation decision could work for us. We don't bundle libuv normally, but many downstream users of libraft (via dqlite) build everything from source, and we could apply a patch in that situation.

**Reply by bnoordhuis** [reply]
For future reference: https://github.com/libuv/libuv/pull/4039

**Reply by cole-miller** [reply]
Thanks bnoordhuis, that's very useful!
